clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,174 @@
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel13>(F); // 10th instantiation throwing error
+  // expected-error@#11{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel14>(F); // 11th instantiation throwing error
+  // Using the same functor twice should be fine
+}
+
+template <typename Ty>
+auto func() -> decltype(__builtin_unique_stable_name(Ty::str));
+
+struct Derp {
+  static constexpr const char str[] = "derp derp derp";
+};
+
+template <typename T>
+static constexpr const char *output1 = __builtin_unique_stable_name(T);
+
+#define MACRO12()                \
+  auto l12 = []() { return 1; }; \
+  constexpr const char *l1_output = __builtin_unique_stable_name(l12);
+
+int main() {
+
+  // kernel0
+  // kernel_single_task<class kernel0>(func<Derp>());
+
+  // kernel1 Evaluate the builtin in a constant context, then call the kernel with that same type
+  auto l1 = []() {};
+  constexpr const char *l1_output = __builtin_unique_stable_name(l1);
+  kernel_single_task<class kernel1>(
+      [=]() { l1(); });
+
+  // kernel2 Call to the builtin on a lambda, then using the lambda for a kernel name
+  auto l2 = []() { return 1; };
+  constexpr const char *l2_output = __builtin_unique_stable_name(l2); // expected-note {{'__builtin_unique_stable_name' evaluated here}}
+  // expected-error@#1{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel2>(l2); // 1st instantiation throwing error
+
+  // kernel3
+  // The current function is named with a lambda (that is, takes a lambda as a template parameter).  Call the builtin (perhaps on the current function?), then pass it to a kernel.
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel3func([]() {});
+
+  // kernel4 Same as above, but you use a NEW lambda defined inline to name the kernel.
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel4func([]() {});
+
+  // kernel7 - expect error
+  // Same as kernel6, except make l6 part of naming kernel
+  auto l7 = []() { return 1; };
+  auto l8 = [](decltype(l7) *derp = nullptr) { return 2; };
+  constexpr const char *l7_output = __builtin_unique_stable_name(l7); // expected-note {{'__builtin_unique_stable_name' evaluated here}}
+  // expected-error@#2{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel7>(l8); // 2nd instantiation throwing error
+
+  // kernel8 and kernel9 - expect no error
```

Comment:
Either the comment is wrong or the test should be failing -- I think the comment is correct though, I would not expect errors here. @erichkeane, ideas?

Reply:
I think this comment is wrong and we do expect these errors here.

Reply:
The latest test output agrees with our original understanding that this test shouldn't emit errors. Just adding this comment here to make sure that we all agree.

Reply:
The latest test output *does* show an error, but the latest code comment also says that we expect the error (so the comment and the test results agree now). I believe we *do* expect the errors in that case, so I think this is all correct.

Reply:
Yes. That makes sense. I should've been clearer: It shows the expected error on only one of the (expected) branches, but not both.